### PR TITLE
Support restricted fields on AccessGuard

### PR DESCRIPTION
**Pull request by belgamo**
related to https://github.com/getjerry/nest-casl/issues/254

### Motivation
Currently, there's no effortless way to support restricted fields by using the `AccessGuard` which requires us to manually call either `hasAbility` or `assertAbility` from `AccessService`.

### Strategy
Get all properties from the body object sent by the client and transform them into dot notation paths, that is the way CASL compares the desired ability against the predefined rules. It's fully compatible with [CASL patterns](https://casl.js.org/v5/en/guide/restricting-fields#field-patterns-table).

### Goal
Deny the request if the user has sent a field in which they don't have permission to perform an action.

### CODE REVIEW

**Comment by belgamo** [comment]
@liquidautumn I'd love to get your input here. Please, review whenever you get a chance.

File: package.json [hunk 1]
```diff
@@ -39,7 +39,8 @@
     "@ucast/core": "^1.10.1",
     "@ucast/mongo": "^2.4.2",
     "@ucast/mongo2js": "^1.3.3",
-    "@ucast/sql": "^1.0.0-alpha.1"
+    "@ucast/sql": "^1.0.0-alpha.1",
+    "dot-object": "^2.1.4"
```
Comment (liquidautumn) [inline package.json:43]: I'm a bit concerned about `dot-object` not being maintained. I did a quick search and found https://github.com/sindresorhus/dot-prop with less dependencies and released recently, could be a drop in replacement

**Comment by liquidautumn** [comment]
> @liquidautumn I'd love to get your input here. Please, review when you get a chance.

Thank you for contribution, it looks good overall.
Commented on dependency issue, let me know what you think. Also docs update needed but not critical, might be done separately.

File: package.json (hunk 1 above)
Comment (belgamo) [reply package.json:43]: I think `dot-prop` doesn't have a plug-and-play solution like `dot-object`. However, I just need the `.dot` parser and it's about [30 lines of code](https://github.com/rhalff/dot-object/blob/master/src/dot-object.js#L516). We can easily extract it from there and make it a utility in nest-casl.

Comment (belgamo) [reply package.json:43]: > I'm a bit concerned about `dot-object` not being maintained. I did a quick search and found https://github.com/sindresorhus/dot-prop with less dependencies and released recently, could be a drop in replacement

I think `dot-prop` doesn't have a plug-and-play solution like `dot-object`. However, I just need the `.dot` parser and it's about [30 lines of code](https://github.com/rhalff/dot-object/blob/master/src/dot-object.js#L516). We can easily extract it from there and make it a utility in nest-casl.

**Comment by belgamo** [comment]
> > @liquidautumn I'd love to get your input here. Please, review when you get a chance.
> 
> Thank you for contribution, it looks good overall. Commented on dependency issue, let me know what you think. Also docs update needed but not critical, might be done separately.

I'm wondering if we really need to document this... Since it's built on top of CASL, I'd expect the [restricting fields](https://casl.js.org/v6/en/guide/restricting-fields) feature to just work when I call `can/cannot` methods while defining permissions. That's my perspective as a consumer of this library. If you really think it's nice to have, we can evidence it with a short example.

File: package.json (hunk 1 above)
Comment (liquidautumn) [reply package.json:43]: https://www.npmjs.com/package/flat should do, it's focused on what needed and neatly maintained

Comment (belgamo) [reply package.json]: > https://www.npmjs.com/package/flat should do, it's focused on what needed and neatly maintained

Yep, that works. However, I had to install v5.0.2 which is the last one built with CommonJS. From v6 on, they use ESM which is not supported by Jest without making custom configs.

**Review by liquidautumn — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (liquidautumn) [reply package.json]: It's ok, I'm going to replace jest with vite as jest blocks another package update, will update this one too.

**Pull request merged**